forwards_paper.tex

```diff
@@ -903,59 +903,80 @@ \subsubsection*{Estimates of run-time complexity}
-One simply needs to fill the relevant columns,
-and to send them to \msprime{} at regular intervals for simplification.
-For example, in the benchmarking simulations above,
-we used the \fwdpp{} \cpp{} API \citep{fwdpp}
-to fill arrays of simple structures representing the node and edge data.
-Doing so requires no modification of the \fwdpp{} code base.
-Rather, we simply need to bookkeep parent/offspring labels,
-and perform simple processing of the recombination breakpoints from each mating event.
-Using \texttt{pybind11} (\url{https://github.com/pybind/pybind11/}),
-the node/edge arrays on the \cpp{} side are made visible, without copy, to Python as a multi-column numpy array,
-which is processed by the \msprime{} tables API.
-Our implementation also uses \texttt{pybind11}
-to create a Python package extending \texttt{fwdpy11}
-(a Python package based on \fwdpp{}; \url{https://github.com/molpopgen/fwdpy11}),
-% (A neat aside, that we may wish to skip here, is that it may be possible to use the tables
-% API from a command-line C/\cpp{} application via an embedded Python interpreter.)
+For small scale data and debugging purposes, a simple text based format is also supported.
+
+The \msprime\ Python Tables API provides a general purpose toolkit for importing
+and processing succinct tree sequences. Interoperation with Python simulators
+is then straightforward. For the proof-of-concept implementations discussed
+here, we took two different approaches to doing this. In the first method,
+based on \simupop{}, the simulation engine invokes callbacks at critical points
+of the simulation, and we infer nodes and edges from the information that is
+provided. Rows are appended to the tables one-by-one, and the tables are
+periodically sorted and simplified to control memory usage. The second method we developed uses
+\texttt{pybind11} (\url{https://github.com/pybind/pybind11/}) to inferface
+with the \fwdpp{} \cpp{} API \citep{fwdpp}. No modifications were
+required to the \fwdpp{} code base; rather, we simply need to bookkeep parent/offspring labels,
+and perform simple processing of the recombination breakpoints from each mating
+event to generate node and edge data. This information is then periodically copied
+to the \msprime\ Tables API, where it is sorted and simplified.
+
+While these methods work well, and indeed result in significant performance
+gains over non-ancestry based simulation, they are far from optimal. Because
+forwards time simulation is highly CPU intensive, most simulation engines
+are written in C or C++ out of necessity. Therefore, the simplest and most
+efficient approach would be to call the \msprime\ C library directly as part of
+the core simulation logic, in the manner outlined in Algorithm W. If designed
+in this way, the simulation engine would not need to keep copies of the node
+and edge information and would not need to infer edges;
```

Comment:
"infer edges" isn't true in the c++ implementation so is a bit confusing to state as a general issue. Should remove as simuPOP details will be in appendix if present at all.

Reply:
I'd have to reread pybind11 doc, but my understanding is that I am not using the numpy api, actually.  I'm using the buffer protocol and registering the stored struct as numpy dtypes.  The conversion to a numpy record arrays is zero copy due to buffer protocol.

Pybind11's entire numpy support works without ever including numpy headers nor linking to it.  Black magic, for sure.